### ISSUE TITLE
@-mentions doing slightly weird things with fediverse usernames

### ISSUE THREAD

**Issue by momijizukamori** [dreamwidth/dreamwidth#3042]
![image](https://user-images.githubusercontent.com/2245253/203387534-07080c29-b0e4-4fad-84c8-ce5eba871896.png)

[This entry](https://momijizukamori.dreamwidth.org/469281.html) was posted as 'Casual HTML' so I was not expecting the (erroneous) username interpolation - the original input text is `@cocoa@thread.seam.rip`. I assume our conversion filter has some logic to not convert email addresses but the leading @ in typical fediverse usernames probably needs to be exempted too.

**Comment by deborahgu** [comment]
@momijizukamori what do you think of the idea of adding fediverse usernames? They'd have to be tagged specifically since the domain names vary so much, but something like `@username@host.name.whatever@fedi`? or `<user name="username" site="host.name.whatever" protocol="fediverse">`?

**Comment by momijizukamori** [comment]
I am in favor! Though as a first step I would like to get it to *not* try to turn fediverse usernames into two different account names (which is what it does if you try to add one while the parser is set to markdown)

One possible option would be to assume anything in the `@user@sitename.com` format that isn't one of the other supported external sites is a fedi site, though I need to investigate if various distributions all share similar user profile addressing schemes. Also trying to figure out if there was a reason we decided on `@momijizukamori.twitter` for example as the shortform instead of `@momijizukamori@twitter.com`

**Comment by momijizukamori** [comment]
Also for my future reference, so I can find it again - this is where mentions get translated: https://github.com/dreamwidth/dreamwidth/blob/ede4d262368b6b0c25ba56b1ea81c423b3261833/cgi-bin/LJ/CleanHTML.pm#L1957

**Comment by deborahgu** [comment]
Yeah, I was also wondering about if there was a reason that the short form was that way – it was so unexpected (especially because I would expect some sites to support dots and usernames, but no sites to support @  symbols in usernames, this is not uucp), that the mere fact that that’s the pattern we use made me think there was some reason for it.

**Comment by momijizukamori** [comment]
Yeah, I'm hoping someone who's poked at that part of the code remembers *why* we did it that way so we don't go through a discovery process the hard way, lol

**Comment by kareila** [comment]
https://dw-suggestions.dreamwidth.org/1527807.html

As far as I can tell, user.sitename and user@sitename were both proposed, and Mark decided to implement the former since there was no prevailing strong argument in favor of one or the other. And @ signs embedded within other strings were ignored in order not to break other things, like email addresses and Medium.com URLs.

What isn't clear to me is why Casual HTML and Markdown would be parsing the usernames differently - it looks like they follow the same code path for doing the conversion.

For now, if all you want to do is make sure `@username@mastodon.tld` doesn't get converted incorrectly, you can always escape the affected @ sign(s) with a backslash.
